lib/Serialization/ASTWriter.cpp

```diff
@@ -286,22 +286,28 @@ void ASTTypeWriter::VisitFunctionProtoType(const FunctionProtoType *T) {
 
   Record.push_back(T->isVariadic());
   Record.push_back(T->hasTrailingReturn());
+  Record.push_back(T->hasParamBounds());
   Record.push_back(T->getTypeQuals());
   Record.push_back(static_cast<unsigned>(T->getRefQualifier()));
   addExceptionSpec(T, Record);
+  Record.AddStmt(const_cast<BoundsExpr *>(T->getReturnBounds()));
 
   Record.push_back(T->getNumParams());
   for (unsigned I = 0, N = T->getNumParams(); I != N; ++I)
     Record.AddTypeRef(T->getParamType(I));
 
+  if (T->hasParamBounds())
+    for (unsigned I = 0, N = T->getNumParams(); I != N; ++I)
+      Record.AddStmt(const_cast<BoundsExpr *>(T->getParamBounds(I)));
+
   if (T->hasExtParameterInfos()) {
     for (unsigned I = 0, N = T->getNumParams(); I != N; ++I)
       Record.push_back(T->getExtParameterInfo(I).getOpaqueValue());
   }
 
-  if (T->isVariadic() || T->hasTrailingReturn() || T->getTypeQuals() ||
-      T->getRefQualifier() || T->getExceptionSpecType() != EST_None ||
-      T->hasExtParameterInfos())
+  if (T->isVariadic() || T->hasTrailingReturn() || T->hasParamBounds() ||
```

Comment:
Yes, this is confusing.  AbbrevToUse indicates whether something should be compressed while a the record out.  There are special "domain-specific" compressions wired into the record writer.   For example, there are domain-specific compressions for types that are existing types with some qualifiers.   There is also one for function prototypes, but it looks it is not enabled.  ASTWriter::ASTWriteTypeAbbrevs  seems to create a template for a function prototype abbreviation.   The template omits a bunch of fields.  The code that you pointed to disables abbreviations for those fields.   Because I didn't add the new fields for Checked C to the template, I make sure that compression will be disabled for them too.

Somewhat confusing,  I couldn't find any place where the function prototype abbreviation is actually being used.  It looks like Abbrev is always set to 0 for function prototypes.  

I'll add a comment to the code explaining the if-statement.